### PR TITLE
Acquiring locks are deferred to commit time

### PR DESCRIPTION
**Pull request by lutovich**
This is a reimplementation of PR #7545 for 3.0 branch.

Deferred locks is a kernel extension and jar is not part of the distribution
but is published to maven.

It is also guarded by `unsupported.dbms.deferred_locks.enabled` setting which
is off by default.

We separate between optimistic and pessimistic locking:

**Optimistic locking**
Acquire locks triggered by some write or read event through the `Statement` cake.
Instead of grabbing those locks we only remind what resources we want to lock
and then acquire the them right before commit.

**Pessimistic locking**
Locks grabbed explicitly by the user or grabbed on master by a slave. Those
locks are taken immediately.

**Note**
Grabbing locks just before commit guards for store inconsistencies but not
logical inconsistencies.

**Warning**
This feature should not be used. Most of all code in Neo4j is assumed to be
executed under some sort of lock, exclusive or shared. This commit makes that
assumption false and there has not been enough investigation to foresee what
consequences that will have for transaction execution. These problems are
still in the unknown but there are also known, purposely undocumented,
problems.

### CODE REVIEW

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/ConstraintEnforcingEntityOperations.java [hunk 1]
```diff
@@ -329,7 +329,7 @@ public long nodeGetFromUniqueIndexSeek(
         }
 
         // If we find the node - hold a shared lock. If we don't find a node - hold an exclusive lock.
-        Locks.Client locks = state.locks();
+        Locks.Client locks = state.locks().optimistic();
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/ConstraintEnforcingEntityOperations.java]: looks like comment on top is no longer valid

File: enterprise/deferred-locks/src/test/java/org/neo4j/kernel/impl/locking/DeferringLocksIT.java [hunk 2]
```diff
@@ -0,0 +1,453 @@
...
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+import static org.neo4j.helpers.collection.Iterables.count;
+
+public class DeferringLocksIT
+{
+    private static final long TEST_TIMEOUT = 30_000;
+
+    private static final Label LABEL = DynamicLabel.label( "label" );
```
Comment (MishaDemianenko) [inline enterprise/deferred-locks/src/test/java/org/neo4j/kernel/impl/locking/DeferringLocksIT.java]: this should be just `Label.label`

File: enterprise/kernel/src/main/java/org/neo4j/kernel/impl/enterprise/lock/forseti/ForsetiClient.java [hunk 3]
```diff
@@ -150,87 +150,91 @@ public void acquireShared( ResourceType resourceType, long resourceId ) throws A
...
-                heldShareLocks.put( resourceId, Math.incrementExact( heldCount ) );
-                return;
-            }
+                // First, check if we already hold this as a shared lock
+                int heldCount = heldShareLocks.get( resourceId );
+                if ( heldCount != -1 )
+                {
+                    // We already have a lock on this, just increment our local reference counter.
+                    heldShareLocks.put( resourceId, Math.incrementExact( heldCount ) );
+                    return;
```
Comment (MishaDemianenko) [inline enterprise/kernel/src/main/java/org/neo4j/kernel/impl/enterprise/lock/forseti/ForsetiClient.java]: do not think we want to return from here

File: enterprise/kernel/src/main/java/org/neo4j/kernel/impl/enterprise/lock/forseti/ForsetiClient.java [hunk 4]
```diff
@@ -150,87 +150,91 @@ public void acquireShared( ResourceType resourceType, long resourceId ) throws A
...
-                return;
-            }
+                // Second, check if we hold it as an exclusive lock
+                if ( heldExclusiveLocks.containsKey( resourceId ) )
+                {
+                    // We already have an exclusive lock, so just leave that in place.
+                    // When the exclusive lock is released, it will be automatically downgraded to a shared lock,
+                    // since we bumped the share lock reference count.
+                    heldShareLocks.put( resourceId, 1 );
+                    return;
```
Comment (MishaDemianenko) [inline enterprise/kernel/src/main/java/org/neo4j/kernel/impl/enterprise/lock/forseti/ForsetiClient.java]: same here, we should not return from here

File: enterprise/kernel/src/main/java/org/neo4j/kernel/impl/enterprise/lock/forseti/ForsetiClient.java [hunk 5]
```diff
@@ -249,40 +253,43 @@ public void acquireExclusive( ResourceType resourceType, long resourceId ) throw
...
-            while ( (existingLock = lockMap.putIfAbsent( resourceId, myExclusiveLock )) != null )
+            for ( long resourceId : resourceIds )
             {
-                assertNotStopped();
+                int heldCount = heldLocks.get( resourceId );
+                if ( heldCount != -1 )
+                {
+                    // We already have a lock on this, just increment our local reference counter.
+                    heldLocks.put( resourceId, Math.incrementExact( heldCount ) );
+                    return;
```
Comment (MishaDemianenko) [inline enterprise/kernel/src/main/java/org/neo4j/kernel/impl/enterprise/lock/forseti/ForsetiClient.java]: same here we should not return here

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/LockingStatementOperations.java [hunk 6]
```diff
@@ -106,9 +104,9 @@ public boolean nodeAddLabel( KernelStatement state, long nodeId, int labelId )
         //
         // It would be cleaner if the schema and data cakes were separated so that the SchemaReadOperations object used
         // by ConstraintEnforcingEntityOperations included the full cake, with locking included.
-        state.locks().acquireShared( ResourceTypes.SCHEMA, schemaResource() );
+        acquireSharedSchemaLock(state);
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/LockingStatementOperations.java]: please reformat some of the updated statements

**Comment by MishaDemianenko** [comment]
@lutovich couple of comments and also rebased required already

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/ConstraintEnforcingEntityOperations.java (hunk 1 above)
Comment (lutovich) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/ConstraintEnforcingEntityOperations.java]: i think it is still valid, all those locks can just be deferred

**Comment by lutovich** [comment]
@MishaDemianenko comments fixed now, added two more commits and rebased.

**Comment by MishaDemianenko** [comment]
@lutovich i think we need to rebase this one to pick up latest build env settings

**Comment by lutovich** [comment]
@MishaDemianenko rebased

**Pull request merged**